### PR TITLE
SONARJAVA-1514 handle invariance of generics

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 1]
```diff
@@ -552,24 +552,20 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
   }
 
   private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing, boolean isParameterized) {
-    // FIXME SONARJAVA-1514 Inference of types for method using parameter types is not handled
-    if (!isParameterized && usesWildcard(formal)) {
+    // FIXME SONARJAVA-1298 type parameters should be handled in formal arguments prior to the call to this method
+    if (!isParameterized && (isParametrizedType(arg) || isParametrizedType(formal) || isWilcard(arg) || isWilcard(formal))) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1067)

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 2]
```diff
@@ -75,11 +76,33 @@ protected void onMethodInvocationFound(MethodInvocationTree tree) {
...
-    if (collectionParameterType != null && !collectionParameterType.isUnknown() && !isArgumentCompatible(argumentType, collectionParameterType)) {
+    if (collectionParameterType != null && !collectionParameterType.isUnknown() && !isArgumentCompatible(argumentType, collectionParameterType)
+      && !useTypeParameters(argumentType)) {
       reportIssue(MethodsHelper.methodName(tree), MessageFormat.format("A \"{0}<{1}>\" cannot contain a \"{2}\"", collectionType, collectionParameterType, argumentType));
     }
   }
 
+  // FIXME SONARJAVA-1298 - method to remove
+  // should correctly resolve type parameters, and consequently correctly handle argument compatibility
+  private boolean useTypeParameters(Type type) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: MAJOR The Cyclomatic Complexity of this method "useTypeParameters" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Make "useTypeParameters" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: java-squid/src/main/java/org/sonar/java/resolve/Types.java [hunk 3]
```diff
@@ -66,15 +66,14 @@ public boolean isSubtype(JavaType t, JavaType s) {
...
-          if(t.tag != s.tag) {
-            result = false;
-            break;
-          }
           result = t.isSubtypeOf(s);
           break;
         case JavaType.BOT:
           result = s.tag == JavaType.BOT || s.tag == JavaType.CLASS || s.tag == JavaType.ARRAY;
           break;
+        case JavaType.WILDCARD:
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Types.java]: MAJOR This case's code block is the same as the block for the case on line 68. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1871)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/00b81e497728c8363a73de362f4d6d6af5ab557d/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L475) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 4]
```diff
@@ -75,11 +76,33 @@ protected void onMethodInvocationFound(MethodInvocationTree tree) {
...
-    if (collectionParameterType != null && !collectionParameterType.isUnknown() && !isArgumentCompatible(argumentType, collectionParameterType)) {
+    if (collectionParameterType != null && !collectionParameterType.isUnknown() && !isArgumentCompatible(argumentType, collectionParameterType)
+      && !(useTypeParameters(argumentType) || useTypeParameters(collectionType))) {
       reportIssue(MethodsHelper.methodName(tree), MessageFormat.format("A \"{0}<{1}>\" cannot contain a \"{2}\"", collectionType, collectionParameterType, argumentType));
     }
   }
 
+  // FIXME SONARJAVA-1298 - method to remove
+  // should correctly resolve type parameters, and consequently correctly handle argument compatibility
+  private static boolean useTypeParameters(Type type) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: MAJOR The Cyclomatic Complexity of this method "useTypeParameters" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java [hunk 5]
```diff
@@ -75,11 +76,33 @@ protected void onMethodInvocationFound(MethodInvocationTree tree) {
     // can be null when using raw types
     Type collectionParameterType = getTypeParameter(collectionType);
 
-    if (collectionParameterType != null && !collectionParameterType.isUnknown() && !isArgumentCompatible(argumentType, collectionParameterType)) {
+    if (collectionParameterType != null && !collectionParameterType.isUnknown() && !isArgumentCompatible(argumentType, collectionParameterType)
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CollectionInappropriateCallsCheck.java]: MAJOR Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1067)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/e0711c769a64b150e74436b54225067739b5196b/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L475) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 6]
```diff
@@ -552,26 +552,43 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
...
     return types.isSubtype(arg.erasure(), formal.erasure()) || (autoboxing && isAcceptableByAutoboxing(arg, formal.erasure()));
   }
 
-  private static boolean usesWildcard(JavaType type) {
-    if (type instanceof JavaType.ParametrizedTypeJavaType) {
-      JavaType.ParametrizedTypeJavaType parametrizedType = (JavaType.ParametrizedTypeJavaType) type;
-      for (JavaType substitution : parametrizedType.typeSubstitution.substitutedTypes()) {
-        if (substitution.isTagged(JavaType.WILDCARD)) {
+  // FIXME SONARJAVA-1298 to be removed
+  private static boolean useTypeVariable(JavaType type) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "useTypeVariable" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 7]
```diff
@@ -552,26 +552,43 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
...
     // fall back to behavior based on erasure
     return types.isSubtype(arg.erasure(), formal.erasure()) || (autoboxing && isAcceptableByAutoboxing(arg, formal.erasure()));
   }
 
-  private static boolean usesWildcard(JavaType type) {
-    if (type instanceof JavaType.ParametrizedTypeJavaType) {
-      JavaType.ParametrizedTypeJavaType parametrizedType = (JavaType.ParametrizedTypeJavaType) type;
-      for (JavaType substitution : parametrizedType.typeSubstitution.substitutedTypes()) {
-        if (substitution.isTagged(JavaType.WILDCARD)) {
+  // FIXME SONARJAVA-1298 to be removed
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 8]
```diff
@@ -552,26 +552,43 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
   }
 
   private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing, boolean isParameterized) {
-    // FIXME SONARJAVA-1514 Inference of types for method using parameter types is not handled
-    if (!isParameterized && usesWildcard(formal)) {
+    // FIXME SONARJAVA-1298 type parameters should be handled in formal arguments prior to the call to this method
+    if (!isParameterized && (isParametrizedType(arg) || isParametrizedType(formal) || isWilcardType(arg) || isWilcardType(formal))
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Reduce the number of conditional operators (6) used in the expression (maximum allowed 3). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1067)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 9]
```diff
@@ -552,26 +552,43 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
   }
 
   private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing, boolean isParameterized) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "isAcceptableType" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* MAJOR 7 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/6354db1aa35455b9aca0ca83787fb8393898ef1e/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L475) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* MAJOR 7 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 12 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/fd3a75c72d59d205dfdfcfe7e3ec9e4cb8577a21/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L475) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MINOR [Define a constant instead of duplicating this literal "java.lang.Object" 4 times.](https://github.com/SonarSource/sonar-java/blob/fd3a75c72d59d205dfdfcfe7e3ec9e4cb8577a21/java-squid/src/main/java/org/sonar/java/resolve/JavaType.java#L100) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1192)

File: java-frontend/src/main/java/org/sonar/java/resolve/JavaType.java [hunk 10]
```diff
@@ -452,9 +462,13 @@ public WildCardType(JavaType bound, BoundType boundType) {
       this.symbol.type = this;
     }
 
+    public JavaType bound() {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/JavaType.java]: Unused method. Will be removed.

File: java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 11]
```diff
@@ -552,24 +552,20 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
   }
 
   private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing, boolean isParameterized) {
-    // FIXME SONARJAVA-1514 Inference of types for method using parameter types is not handled
-    if (!isParameterized && usesWildcard(formal)) {
+    // FIXME SONARJAVA-1298 type parameters should be handled in formal arguments prior to the call to this method
+    if (!isParameterized && (isParametrizedType(arg) || isParametrizedType(formal) || isWilcardType(arg) || isWilcardType(formal))) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1067)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 12 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/8fd79abf518208c21732a21ec711088363f918e5/java-frontend/src/main/java/org/sonar/java/resolve/JavaType.java#L475) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MINOR [Define a constant instead of duplicating this literal "java.lang.Object" 4 times.](https://github.com/SonarSource/sonar-java/blob/8fd79abf518208c21732a21ec711088363f918e5/java-frontend/src/main/java/org/sonar/java/resolve/JavaType.java#L100) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1192)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 12 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/39e4d34cfab77490824cb58027168ab491586d26/java-frontend/src/main/java/org/sonar/java/resolve/JavaType.java#L471) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MINOR [Define a constant instead of duplicating this literal "java.lang.Object" 4 times.](https://github.com/SonarSource/sonar-java/blob/39e4d34cfab77490824cb58027168ab491586d26/java-frontend/src/main/java/org/sonar/java/resolve/JavaType.java#L100) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1192)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isSubtypeOf" is 12 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/39e4d34cfab77490824cb58027168ab491586d26/java-frontend/src/main/java/org/sonar/java/resolve/JavaType.java#L471) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MINOR [Define a constant instead of duplicating this literal "java.lang.Object" 4 times.](https://github.com/SonarSource/sonar-java/blob/39e4d34cfab77490824cb58027168ab491586d26/java-frontend/src/main/java/org/sonar/java/resolve/JavaType.java#L100) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1192)

**Pull request merged**